ICE in retrieving ambiguous associated type - `E0223`

## ICE in retrieving ambiguous associated type - [`E0223` ](https://doc.rust-lang.org/error_codes/E0223.html#error-code-e0223)


- You can view the same on [`compiler-explorer`](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATKUcAZPAZMADl3ACNMYhAADlIAB1QFQjsGFzcPb3jE5IEAoNCWCKjYq0wbFKECJmICNPdPH1LygUrqgjyQ8MiYyyqauozGvvbAzsLu6IBKS1QTYmR2DgBSLwBmMDAlgFYAITFaVAB3CBMGM0x0Se2AESWNAEEqBgBqFiZAiEnnpYB2Hbv7gRiO8CM8ABKYWgHb5/AHPeHPAgATzimGeAHUEKglqt/g84Qins8EJCDp9ngBaHHYb6rJx4VZeakgEAATVmOJ2zwA9NznpFiCQNgDfrd8T9rhxprROFteJ5uLxUJwAEpmUEKWbzNErVY8UgETRS6YAawk0QAdD8fhoAGxbLhrDQaACceq2sRlHEk8qNpGVHF4ChAGgNRumcFgMEQKFQLDidEi5EoaHjiaiwGQyGICgpCmYcQUWIsNFoBEiwYgYT9YUC1SRnH1teYxCRAHkwtoyobFaRU2xBG2GLQGxwtKQsGETMAnPtg72sG8jOIx7x8MRu3gAG6YefjzCqMomcuN3iBcte8e0PBhYGtlxYP1AvAsU%2BkHfEMKJTDXTBL4DXkY4Z8AYwAKAAangmCHG2qIKvq/CCCIYjsFIMiCIoKjqKupC6Fw%2BiGMY6r6DewaQNMqBxLYAjzkqH7EHgWBkVAzBsCAFw5Aw75iCYixeBoXg8JM0xNNRngQI4AyePh/ijAURR6AkSRiVJinZGJHTyd0%2BGiRUwyqTpkKbi0wyaV0UQ6fprj1Ho5htGZ4wWSJWoLBI0qyr6OEBs8qjRLaFK2pIzwAOJOE4KpCM8ECheFkVvOYkRfBAuCECQ3xrFwky8D2WjCaQZpbKGXo%2BqQr6SC6FrRFstqCVwtqrLaGhVS6tqkAq44BkGIZhquEbRhASCzAQcTHsmECpgm9DEMErCLKY5jIM8XAWralr7vgRAMegeiIcIojiGhu2YWofp4aQhzAnEp7uRwcptX6AZtseI2gqgVA%2BX5AVBTFEVRT9cVMAlxBJS4aZTelepZT1uU3SVr6FRaUirD80TRA1qwulIGjSO1SqcF1oY5ca%2BXeD8lWY41Ug/FsNpcNEGhbPonCrJ5HX49DUqkJGUD9UgE3pmN/NTSAxBSC6oaluWOaUNWOHNvWb7y62HZdjYb79owBBDiOfqTtOs5QvO%2BqLoRK7juum47nuvAHkeJ69uekJ%2Btet71g%2Bizjs%2Br69h%2BX5KL%2B/6AaAvUgUwYGQdBsGMG%2Bu3IQd0hHUoJ04boPgGEB80WC7zEUVRKTzhSFLZkw5YUsiqI4tczZ0QKjG7vAIlGc04mSdZGQyQw6AOQp%2BFKZxBlZMpKTd9plhN2JrT9G30lj9YE%2BmXJ5m2VZ6Qz3ZNQj05MxzK5mVM7drN4xwzyZ4ty3lRaGhRSlm0Q5l2XhtMJJMFgUSfCTXDOvvcMgIV91eezYMhNgLcwGiAIaL1BZxkmpEGabEOC%2BX8oFEKYVfrRVQQDIGUMOKbUYjtWQcdUIJ1kMdbC45U7nUutdfed1cb%2Bk4E9Yax5nhvRPuqM%2BFoL5XwgKDGBxAIZeChkTPKz9X6UFNBIL%2BxVeDw1DHQzqlhurCIkWsK0boUY2i8F4F0WwtGtS9Czf%2BbNAwcz6vAUBfNoECwoONKxwtRY/HFnwOgUtKyy3HErUcTY6zK07N2dWcYBxa2HKOc2mApwzjnG%2BE2y4PZrjwBucoVs/S22QMeRY%2BpHaXl4C7O8SJ3ZPgYt7fUvtvwB1NoEYOuVQ7hygjBOCMcCH7SIeheQScyE6BAKsAiGdiLZwbv6PONF2aoHonXHOs9jIt07gPWS%2BQl693UikAefcNKL0crZceek2gD10iZey6ye69B2dPZeBz5kbL3pqHeqEbq0IepwNhC0lpqMvtfDaaVdT31MU/TAL9ujvzNF4SQFothbEkBjBm9VqqSGiC6T0nAf5/3kYApRIDebgOeqNGxQtYGzU4KfF5XCbYfK2vgpCzSJDEIwu006XTKFMCuoqO5h96EcEYS9Fh71CXLUcW8nhdjIgQ1WEIx%2BpBREAthjI3%2BciHkmKAT8km2jQUQpagJH4jpdHRHwgY1lCjlH7y8Hq9mBqPxJHsJIIAA%3D%3D)
- Previous Work https://github.com/Rust-GCC/gccrs/issues/843

---

### Code

```rust
#![allow(unused)]
fn main() {
trait Hello {
    type Who;

    fn hello() -> <i32>::You; // error[E0223]: ambiguous associated type
}
}
```

---

### Meta

- What version of Rust GCC were you using, git sha d67405831843e84a419be52743fe8bf43dc9d9ae
- gccrs (Compiler-Explorer-Build-gcc-d67405831843e84a419be52743fe8bf43dc9d9ae-binutils-2.42) 14.0.1 20240309 (experimental)

---

### Error output
- Internal Compiler Error.


### Backtrace
  
```rust
  
➜  gccrs-build gcc/crab1 /home/mhd/Desktop/gccrs/gcc/testsuite/rust/external/t.rs                  
crab1: internal compiler error: in get_as_type_path, at rust/ast/rust-path.h:1220
0x13992ce Rust::AST::QualifiedPathType::get_as_type_path()
        ../../gccrs/gcc/rust/ast/rust-path.h:1220
0x157cac2 Rust::HIR::ASTLowerQualifiedPathInType::visit(Rust::AST::QualifiedPathInType&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-type.cc:148
0x13803dd Rust::AST::QualifiedPathInType::accept_vis(Rust::AST::ASTVisitor&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/ast/rust-path.cc:356
0x157c9d2 Rust::HIR::ASTLowerQualifiedPathInType::translate(Rust::AST::QualifiedPathInType&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-type.cc:131
0x157d5e0 Rust::HIR::ASTLoweringType::visit(Rust::AST::QualifiedPathInType&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-type.cc:287
0x13803dd Rust::AST::QualifiedPathInType::accept_vis(Rust::AST::ASTVisitor&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/ast/rust-path.cc:356
0x157c3c2 Rust::HIR::ASTLoweringType::translate(Rust::AST::Type&, bool)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-type.cc:185
0x157331e Rust::HIR::ASTLowerTraitItem::visit(Rust::AST::Function&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-implitem.cc:243
0x12a59f0 Rust::AST::Function::accept_vis(Rust::AST::ASTVisitor&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/ast/rust-ast.cc:4727
0x1573182 Rust::HIR::ASTLowerTraitItem::translate(Rust::AST::AssociatedItem&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-implitem.cc:212
0x156b43f Rust::HIR::ASTLoweringItem::visit(Rust::AST::Trait&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-item.cc:593
0x12a5d10 Rust::AST::Trait::accept_vis(Rust::AST::ASTVisitor&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/ast/rust-ast.cc:4811
0x1567b62 Rust::HIR::ASTLoweringItem::translate(Rust::AST::Item&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-item.cc:38
0x1581780 Rust::HIR::ASTLoweringStmt::visit(Rust::AST::Trait&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-stmt.cc:145
0x12a5d10 Rust::AST::Trait::accept_vis(Rust::AST::ASTVisitor&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/ast/rust-ast.cc:4811
0x1581106 Rust::HIR::ASTLoweringStmt::translate(Rust::AST::Stmt*, bool*)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-stmt.cc:32
0x1557207 Rust::HIR::ASTLoweringBlock::visit(Rust::AST::BlockExpr&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower.cc:116
0x12a53c0 Rust::AST::BlockExpr::accept_vis(Rust::AST::ASTVisitor&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/ast/rust-ast.cc:4529
0x155935f Rust::HIR::ASTLoweringBlock::translate(Rust::AST::BlockExpr&, bool*)
        ../../gccrs/gcc/rust/hir/rust-ast-lower-block.h:37
0x156a4c4 Rust::HIR::ASTLoweringItem::visit(Rust::AST::Function&)
        /home/mhd/Desktop/gccrs-build/gcc/../../gccrs/gcc/rust/hir/rust-ast-lower-item.cc:443
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```
  

---

--- a/gcc/rust/hir/rust-ast-lower-type.cc
+++ b/gcc/rust/hir/rust-ast-lower-type.cc
@@ -145,8 +145,15 @@ ASTLowerQualifiedPathInType::visit (AST::QualifiedPathInType &path)
 
   HIR::Type *qual_type
     = ASTLoweringType::translate (path.get_qualified_path_type ().get_type ());
-  HIR::TypePath *qual_trait = ASTLowerTypePath::translate (
-    path.get_qualified_path_type ().get_as_type_path ());
+
+  HIR::TypePath *qual_trait = nullptr;
+  if (!path.get_qualified_path_type ().is_error ())
+    {
+      AST::QualifiedPathType &qualifier = path.get_qualified_path_type ();
+      if (qualifier.has_as_clause ())
+	qual_trait
+	  = ASTLowerTypePath::translate (qualifier.get_as_type_path ());
+    }
 
   HIR::QualifiedPathType qual_path_type (
     qual_mappings, std::unique_ptr<HIR::Type> (qual_type),

--- a/gcc/testsuite/rust/compile/issue-3082.rs
+++ b/gcc/testsuite/rust/compile/issue-3082.rs
@@ -1,0 +1,9 @@
+#![allow(unused)]
+fn main() {
+    trait Hello {
+        type Who;
+
+        fn hello() -> <i32>::You;
+        // { dg-error "failed to resolve return type" "" { target *-*-* } .-1 }
+    }
+}

--- a/gcc/testsuite/rust/compile/nr2/exclude
+++ b/gcc/testsuite/rust/compile/nr2/exclude
@@ -277,3 +277,4 @@ dropck_eyepatch_feature_gate.rs
 inline_asm_parse_output_operand.rs
 issue-3030.rs
 issue-3035.rs
+issue-3082.rs

